src/passes/ConstantFieldPropagation.cpp

```diff
@@ -113,22 +151,196 @@ struct FunctionOptimizer : public WalkerPass<PostWalker<FunctionOptimizer>> {
+      SmallVector<HeapType, 10> types;
+
+      // Whether this slot is used. If so, |constant| has a value, and |types|
+      // is not empty.
+      bool used() const {
+        if (constant.hasNoted()) {
+          assert(!types.empty());
+          return true;
+        }
+        assert(types.empty());
+        return false;
+      }
+    } values[2];
+
+    // Handle one of the subtypes of the relevant type. We check what value it
+    // has for the field, and update |values|. If we hit a problem, we mark us
+    // as having failed.
+    auto fail = false;
+    auto handleType = [&](HeapType type, Index depth) {
+      if (fail) {
+        // TODO: Add a mechanism to halt |iterSubTypes| in the middle, as once
+        //       we fail there is no point to further iterating.
+        return;
+      }
+
+      auto iter = rawNewInfos.find(type);
+      if (iter == rawNewInfos.end()) {
+        // This type has no struct.news, so we can ignore it: it is abstract.
+        return;
+      }
+
+      auto value = iter->second[curr->index];
+      if (!value.isConstant()) {
+        // The value here is not constant, so give up entirely.
+        fail = true;
+        return;
+      }
+
+      // Consider the constant value compared to previous ones.
+      for (Index i = 0; i < 2; i++) {
+        if (!values[i].used()) {
+          // There is nothing in this slot: place this value there.
+          values[i].constant = value;
+          values[i].types.push_back(type);
+          break;
+        }
+
+        // There is something in this slot. If we have the same value, append.
+        if (values[i].constant == value) {
+          values[i].types.push_back(type);
+          break;
+        }
+
+        // Otherwise, this value is different than values[i], which is fine:
+        // we can add it as the second value in the next loop iteration - at
+        // least, we can do that if there is another iteration: If it's already
+        // the last, we've failed to find only two values.
+        if (i == 1) {
+          fail = true;
+          return;
+        }
+      }
+    };
+    subTypes.iterSubTypes(refHeapType, handleType);
+
+    if (fail) {
+      return;
+    }
+
+    // If we filled slot 1 then we must have filled slot 0 before.
+    assert(values[1].used() ? values[0].used() : true);
+
+    if (!values[1].used()) {
+      // We did not see two constant values (we might have seen just one, or
+      // even no constant values at all).
+      return;
+    }
+
+    // We have exactly two values to pick between. We can pick between those
+    // values using a single ref.test if the two sets of types are actually
+    // disjoint. In general we could compute the LUB of each set and see if it
+    // overlaps with the other, but for efficiency we only want to do this
+    // optimization if the type we test on is closed/final, since ref.test on a
+    // final type can be fairly fast (perhaps constant time). We therefore look
```

Comment:
I didn't measure it, I was relying on what I have heard about the cost of `ref.test` in general. I'll ask for more clarification.

It would be more complex, however, to do the general thing, I had that code earlier at some point. We need to compute two LUBs and see one is nested in the other, iirc, something like that.